File: qc-core/src/main/java/com/eager/questioncloud/core/domain/hub/board/service/PostService.java
package com.eager.questioncloud.core.domain.hub.board.service;

import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.core.domain.hub.board.dto.PostDto.PostDetail;
import com.eager.questioncloud.core.domain.hub.board.dto.PostDto.PostListItem;
import com.eager.questioncloud.core.domain.hub.board.implement.PostAppender;
import com.eager.questioncloud.core.domain.hub.board.implement.PostReader;
import com.eager.questioncloud.core.domain.hub.board.implement.PostRemover;
import com.eager.questioncloud.core.domain.hub.board.implement.PostUpdater;
import com.eager.questioncloud.core.domain.hub.board.model.Post;
import com.eager.questioncloud.core.domain.hub.board.vo.PostContent;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class PostService {
    private final PostAppender postAppender;
    private final PostReader postReader;
    private final PostUpdater postUpdater;
    private final PostRemover postRemover;

    public Post register(Post post) {
        return postAppender.append(post);
    }

    public List<PostListItem> getPostList(Long userId, Long questionId, PagingInformation pagingInformation) {
        return postReader.getPosts(userId, questionId, pagingInformation);
    }

    public int countPost(Long questionId) {
        return postReader.count(questionId);
    }

    public PostDetail getPostDetail(Long userId, Long postId) {
        return postReader.getPostDetail(userId, postId);
    }

    public void modify(Long postId, Long userId, PostContent postContent) {
        postUpdater.updatePost(postId, userId, postContent);
    }

    public void delete(Long postId, Long userId) {
        postRemover.delete(postId, userId);
    }
}

File: qc-core/qc-post/qc-post-api/src/main/kotlin/com/eager/questioncloud/post/service/PostService.kt
package com.eager.questioncloud.post.service

import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.post.command.DeletePostCommand
import com.eager.questioncloud.post.command.ModifyPostCommand
import com.eager.questioncloud.post.command.RegisterPostCommand
import com.eager.questioncloud.post.domain.Post
import com.eager.questioncloud.post.dto.PostDetail
import com.eager.questioncloud.post.dto.PostPreview
import com.eager.questioncloud.post.implement.*
import org.springframework.stereotype.Service

@Service
class PostService(
    private val postPermissionValidator: PostPermissionValidator,
    private val postRegister: PostRegister,
    private val postReader: PostReader,
    private val postUpdater: PostUpdater,
    private val postRemover: PostRemover
) {
    fun getPostPreviews(userId: Long, questionId: Long, pagingInformation: PagingInformation): List<PostPreview> {
        postPermissionValidator.validatePostPermission(userId, questionId)
        return postReader.getPostPreviews(userId, questionId, pagingInformation)
    }
    
    fun countPost(questionId: Long): Int {
        return postReader.countPost(questionId)
    }
    
    fun getPostDetail(userId: Long, postId: Long): PostDetail {
        val postDetail = postReader.getPostDetail(userId, postId)
        postPermissionValidator.validatePostPermission(userId, postDetail.questionId)
        return postDetail
    }
    
    fun register(command: RegisterPostCommand): Post {
        postPermissionValidator.validatePostPermission(command.writerId, command.questionId)
        return postRegister.register(command)
    }
    
    fun modify(command: ModifyPostCommand) {
        postUpdater.modify(command)
    }
    
    fun delete(command: DeletePostCommand) {
        postRemover.delete(command)
    }
}
